src/controller/python/chip/clusters/Command.py

```diff
@@ -117,20 +128,91 @@ def handleError(self, status: Status, chipError: PyChipError):
             self._handleError, status, chipError, None
         )
 
+    def handleDone(self):
+        ctypes.pythonapi.Py_DecRef(ctypes.py_object(self))
+
+
+class AsyncBatchCommandsTransaction:
+    def __init__(self, future: Future, eventLoop, expectTypes: List[Type]):
+        self._event_loop = eventLoop
+        self._future = future
+        self._expect_types = expectTypes
+        default_im_failure = chip.interaction_model.InteractionModelError(
+            chip.interaction_model.Status.Failure)
+        self._responses = [default_im_failure] * len(expectTypes)
+
+    def _handleResponse(self, path: CommandPath, index: int, status: Status, response: bytes):
+        if index > len(self._responses):
+            self._handleError(status, 0, IndexError(f"CommandSenderCallback has given us an unexpected index value {index}"))
+            return
+
+        if (len(response) == 0):
+            self._responses[index] = None
+        else:
+            # If a type hasn't been assigned, let's auto-deduce it.
+            if (self._expect_types[index] is None):
+                self._expect_types[index] = FindCommandClusterObject(False, path)
+
+            if self._expect_types[index]:
+                try:
+                    self._responses[index] = self._expect_types[index].FromTLV(response)
+                except Exception as ex:
+                    self._handleError(
+                        status, 0, ex)
+            else:
+                self._responses[index] = None
+
+    def handleResponse(self, path: CommandPath, index: int, status: Status, response: bytes):
+        self._event_loop.call_soon_threadsafe(
+            self._handleResponse, path, index, status, response)
+
+    def _handleError(self, imError: Status, chipError: PyChipError, exception: Exception):
+        if self._future.done():
+            # TODO Right now this even happens if there was a legit IM Status error on one command. We need to update OnError to
+            # allow providing a CommandRef that we can try associating with it.
+            logger.exception(f"Recieved another error, but we have sent error. imError:{imError}, chipError {chipError}")
+            return
+        if exception:
+            self._future.set_exception(exception)
+        elif chipError != 0:
+            self._future.set_exception(chipError.to_exception())
+        else:
+            try:
+                self._future.set_exception(
+                    chip.interaction_model.InteractionModelError(chip.interaction_model.Status(imError.IMStatus), imError.ClusterStatus))
+            except Exception:
```

Comment:
This is too wide a net, and can catch other exceptions. What exceptions can set_exception or InteractionModelError constructor raise?

Reply:
This is a copy and paste from `AsyncCommandTransaction _handleError`.

Now that you ask it honestly doesn't make any sense:
`set_exception` based on documentation there `Raises a InvalidStateError error if the Future is already done.`. but I don't think that is the exception they were trying to catch as in the exception they are calling it again.

Again InteractionModelError doesn't raise any exceptions. My guess is they are trying to catch an `AttributeError`, or something else from trying to access members of `imError`.

I use that for now and if things blow up we can have a comment here saying that you found out the exception we were expecting please add here

Reply:
I just think it is better to address these all at one in a separate PR. But it currently hasn't been broken

Reply:
Ops this issue has now actually been addressed :)